src/main/kotlin/eu/europa/ec/eudi/pidissuer/adapter/out/did/CreateDidUrl.kt

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright (c) 2023 European Commission
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package eu.europa.ec.eudi.pidissuer.adapter.out.did
+
+import arrow.core.raise.result
+import com.nimbusds.jose.jwk.Curve
+import com.nimbusds.jose.jwk.ECKey
+import com.nimbusds.jose.jwk.JWK
+import com.nimbusds.jose.jwk.OctetKeyPair
+import com.nimbusds.jose.jwk.RSAKey
+import com.nimbusds.jose.util.Base64URL
+import foundation.identity.did.DIDURL
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
+import org.bouncycastle.jcajce.provider.asymmetric.util.EC5Util
+import org.erwinkok.multiformat.multibase.Multibase
+import org.erwinkok.multiformat.multicodec.Multicodec
+import org.erwinkok.multiformat.util.writeUnsignedVarInt
+import java.io.ByteArrayOutputStream
+
+sealed interface DidMethod {
+    data object KEY : DidMethod
+    data object JWK : DidMethod
+}
+
+/**
+ * Given a [JWK] key it creates a DIDUrl of did methods
+ *
+ * * did:key
+ * * did:jwk
+ *
+ */
+fun createDidUrl(publicKey: JWK, method: DidMethod): Result<DIDURL> = result {
+    when (method) {
+        DidMethod.KEY -> createDidKey(publicKey)
+        DidMethod.JWK -> createDidJwk(publicKey)
+    }
+}
+
+private val supportedEcKeyCurves = setOf(
+    Curve.X25519,
+    Curve.Ed25519,
+    Curve.SECP256K1,
+    Curve.P_256,
+    Curve.P_384,
+    Curve.P_521,
+)
+
+/**
+ * did:key:MULTIBASE(base58-btc, MULTICODEC(public-key-type, raw-public-key-bytes))
+ */
+private fun createDidKey(key: JWK): DIDURL = when (key) {
```

Comment:
The factory methods that are currently private, have a recognizable value, on their own.

- Consider exposing them public and perhaps remove both the sealed interface `DidMethod` and the fun `createDidUrl`